### ISSUE TITLE
[feature request] add keras.fit parameters as attributes to BaseWrapper

### ISSUE THREAD

**Issue by stsievert** [adriangb/scikeras#30]
The keras.fit signature only has 18 parameters. Why not include them as parameters to BaseWrapper.__init__? They can already be passed as `model.fit` keyword args.

``` python
from sklearn.base import BaseEstimator

class BaseWrapper(BaseEstimator):
    def __init__(build_fn=None, verbose=1, callbacks=None, ...):
        self.verbose = verbose
        self.callbacks = callbacks
        ...
```

This implementation would (eventually) help enhance the documentation.

**Comment by stsievert** [comment]
From https://github.com/dask/dask-ml/pull/713#issuecomment-666759169:

>  Some internal parts of Scikit-Learn may inspect the __init__ signature of the model instance and get upset if they find an attribute (ex: verbose) that was set during instantiation but was not part of the __intit_ signature of BaseWrapper.

Why doesn't wouldn't the implementation above resolve that issue?

**Comment by adriangb** [comment]
I think this is generally a good idea! This is closer to what skorch does I think.

Regarding Scikit-Learn compatibility, I do think it would solve the above problem, but I'll have to test.

I do think there are several important consideration:
1) Would this break the existing API (either the `build_fn` based one or the subclassing)?
2) Does this get us closer or further from `sklearn.base`? I always felt that ideally these wrappers should be as close as possible in functionality to the base classes/mixins in `sklearn.base`. But perhaps this was misguided and being close to whatever skorch does makes more sense.
3) Would this method still be able to pass all of Scikit-Learn's estimator checks?

And finally, not a deal breaker, but I think it's nice to be able to cleanly create an equivalent to Scikit-Learn's `MLPRegressor` and `MLPClassifier` classes with just a couple of lines of code.

I'll definitely give this a shot some time in the near future (if you don't get to it first) and see how big of a change this would be and how it plays into the above considerations.

**Comment by adriangb** [comment]
Thinking about this a bit more, I really like this approach, especially because it mirrors what Skorch does.

That said, I think this change only really makes sense if it completely replaces the `**skparams` interface (although the two could technically co-exist, I think it would be ugly and hard to test). That interface is carried over from the old wrappers, so this would completely break backwards compatibility.

So regarding the above concerns I raised, I feel that:

1. This is a big issue since this would be a breaking change.
2. This is a wash. I think this gets us further from `sklearn.base` but closer to Skorch.
3. I think these would all be able to pass.
4. Haven't looked into this yet. This could be a dealbreaker if for example we hardcode arguments for `Sequential` but the user is using a subclassed `Model`. I guess we could keep the current dynamic signature checks to avoid this.

**Comment by stsievert** [comment]
This would be a breaking API change only for this library and wouldn't break any one of the expected behavior from `tf.keras.wrappers.scikit_learn`, correct?

**Comment by adriangb** [comment]
Yes, but the idea is that all `tf.keras.wrappers.scikit_learn` users will migrate to this library once this library is stable. Having to re-write their code would be a huge barrier to this.

**Comment by stsievert** [comment]
How would this be backwards incompatible? All the tests in #22 passed (except linting), including the new tests.

**Comment by adriangb** [comment]
My main concern was capturing all of the parameters. Now that I'm actually looking at the docs for `tf.keras.Model` and `tf.keras.Sequential`, that actually looks doable, and I had not looked at the test results for #22, maybe I should have looked at that to start with.

**Comment by adriangb** [comment]
Sorry if I am a bit slow in understanding your proposals @stsievert, but could you please clarify if your proposal was to add the hardcoded parameters _and_ keep `**sk_params`, or to replace `**sk_params` with the hardcoded parameters?

Going back to the original issue that this is trying to resolve (in https://github.com/adriangb/scikeras/issues/30#issuecomment-666770597), this would only be achieved if all parameters are hardcoded (even those passed to `build_fn`). Otherwise, the following will not comply with the Scikit-Learn API:

```python3
wrappers.KerasClassifier(
    build_fn=...,
    hidden_dim=[100, 20, 0],   # this parameter would be set but not be in the signature, causing issues
    batch_size=BATCH_SIZE,  # this parameter would be okay if we hardcode
)
```

That's not to say that hardcoding the parameters is not valuable in and of itself, but I don't think it would completely solve the issue.

**Comment by stsievert** [comment]
I misunderstood which issue this was. Disregard https://github.com/adriangb/scikeras/issues/30#issuecomment-668627479.

**Comment by adriangb** [comment]
Partially addressed in #47, with some more work to come from future changes to `fit` and parameter routing, as noted in https://github.com/adriangb/scikeras/pull/47#issue-466787237
